Ops/src/OpsWorkerGenetic.cpp

```diff
@@ -0,0 +1,231 @@
+
+ Copyright 2017 Hopsan Group
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+
+ The full license is available in the file LICENSE.
+ For details about the 'Hopsan Group' or information about Authors and
+ Contributors see the HOPSANGROUP and AUTHORS files that are located in
+ the Hopsan source code root directory.
+
+-----------------------------------------------------------------------------*/
+
+//!
+//! @file   OpsWorkerGenetic.h
+//! @author Robert Braun <robert.braun@liu.se>
+//! @date   2018-02-05
+//!
+//! @brief Contains the optimization worker class for the genetic algorithm
+//!
+//$Id$
+
+#include "OpsWorkerGenetic.h"
+#include "OpsEvaluator.h"
+#include "OpsMessageHandler.h"
+#include <math.h>
+#include <random>
+
+using namespace Ops;
+
+//! @brief Initializes a particle swarm optimization
+WorkerGenetic::WorkerGenetic(Evaluator *pEvaluator, MessageHandler *pMessageHandler)
+    : Worker(pEvaluator, pMessageHandler) { }
+
+AlgorithmT WorkerGenetic::getAlgorithm()
+{
+    return Genetic;
+}
+
+
+//! @brief Initialization function
+void WorkerGenetic::initialize()
+{
+    Worker::initialize();
+}
+
+
+//! @brief Executes a particle swarm algorithm. optParticleInit() must be called before this one.
+void WorkerGenetic::run()
+{
+    mpMessageHandler->printMessage("Running optimization with genetic algorithm.");
+
+    distributePoints();
+
+    //Evaluate initial objective values
+    mpEvaluator->evaluateAllPoints();
+    mpMessageHandler->objectivesChanged();
+
+
+    mIterationCounter=0;
+    for(; mIterationCounter<mnMaxIterations && !mpMessageHandler->aborted(); ++mIterationCounter)
+    {
+        for(size_t i=0; i<mNumCandidates; i+=2) {
+            selectParents();
+            crossOver();
+        }
+        mutate();
+
+        mpMessageHandler->candidatesChanged();
+
+        //Evaluate objective values
+        mpEvaluator->evaluateAllCandidates();
+
+
+        for(int i=0; i<mNumPoints; ++i)
+        {
+            calculateBestAndWorstId();
+            int worst = getWorstId();
+            if(mCandidateObjectives[i] < mObjectives[worst])
+            {
+                mObjectives[worst] = mCandidateObjectives[i];
+                mPoints[worst] = mCandidatePoints[i];
+            }
+        }
+
+
+
+        mpMessageHandler->objectivesChanged();
+
+        //Check convergence
+        if(checkForConvergence()) break;      //Use complex method, it's the same principle
+
```

Comment:
I am not sure, but if we break here wont the stepCompleted message be one counter behind?